tests/longevity/longevity_test.go

```diff
@@ -0,0 +1,342 @@
+	chaosMap map[string]int
+	// Stores mapping between chaos level and its freq. Values are hardcoded
+	triggerInterval map[string]map[int]time.Duration
+	// Stores which are disruptive triggers. When disruptive triggers are happening in test,
+	// other triggers are allowed to happen only after existing triggers are completed.
+	disruptiveTriggers map[string]bool
+)
+
+func TestLongevity(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	var specReporters []Reporter
+	junitReporter := reporters.NewJUnitReporter("/testresults/junit_basic.xml")
+	specReporters = append(specReporters, junitReporter)
+	RunSpecsWithDefaultAndCustomReporters(t, "Torpedo : Longevity", specReporters)
+}
+
+var _ = BeforeSuite(func() {
+	InitInstance()
+	populateIntervals()
+	populateDisruptiveTriggers()
+})
+
+var _ = Describe("{Longevity}", func() {
+	var contexts []*scheduler.Context
+	var triggerLock sync.Mutex
+	triggerEventsChan := make(chan *EventRecord, 100)
+	triggerFunctions := map[string]func([]*scheduler.Context, *chan *EventRecord){
+		RebootNode:       TriggerRebootNodes,
+		RestartVolDriver: TriggerRestartVolDriver,
+		CrashVolDriver:   TriggerCrashVolDriver,
+		HAUpdate:         TriggerHAUpdate,
+		EmailReporter:    TriggerEmailReporter,
+		AppTaskDown:      TriggerAppTaskDown,
+	}
+	It("has to schedule app and introduce test triggers", func() {
+		Step(fmt.Sprintf("Start watch on K8S configMap [%s] in namespace [%s]",
+			testTriggersConfigMap, configMapNS), func() {
+			err := watchConfigMap()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		Step("Deploy applications", func() {
+			for i := 0; i < Inst().ScaleFactor; i++ {
+				contexts = append(contexts, ScheduleApplications(fmt.Sprintf("longevity-%d", i))...)
+			}
+			ValidateApplications(contexts)
+		})
+
+		var wg sync.WaitGroup
+		Step("Register test triggers", func() {
+			for triggerType, triggerFunc := range triggerFunctions {
+				go testTrigger(&wg, contexts, triggerType, triggerFunc, &triggerLock, &triggerEventsChan)
+				wg.Add(1)
+			}
+		})
+		CollectEventRecords(&triggerEventsChan)
+		wg.Wait()
+		Step("teardown all apps", func() {
+			for _, ctx := range contexts {
+				TearDownContext(ctx, nil)
+			}
+		})
+	})
+	JustAfterEach(func() {
+		AfterEachTest(contexts)
+	})
+})
+
+func testTrigger(wg *sync.WaitGroup,
+	contexts []*scheduler.Context,
+	triggerType string,
+	triggerFunc func([]*scheduler.Context, *chan *EventRecord),
+	triggerLoc *sync.Mutex,
+	triggerEventsChan *chan *EventRecord) {
+	defer wg.Done()
+
+	minRunTime := Inst().MinRunTimeMins
+	timeout := (minRunTime) * 60
+
+	start := time.Now().Local()
+	lastInvocationTime := start
+
+	for {
+		// if timeout is 0, run indefinitely
+		if timeout != 0 && int(time.Since(start).Seconds()) > timeout {
+			break
+		}
+
+		// Get next interval of when trigger should happen
+		// This intercal can dynamically change by editing configMap
+		waitTime, isTriggerEnabled := isTriggerEnabled(triggerType)
+
+		if isTriggerEnabled {
```

Comment:
can be combined and changed to `if isTriggerEnabled && time.Since(lastInvocationTime) > time.Duration(waitTime) {